app/models/alchemy/page/page_elements.rb

```diff
@@ -6,16 +6,32 @@ module Page::PageElements
     included do
       attr_accessor :do_not_autogenerate
 
-      has_many :elements, -> { order(:position) }
-      has_many :contents, :through => :elements
-      has_and_belongs_to_many :to_be_sweeped_elements, -> { uniq },
+      has_many :elements, -> { where(parent_element_id: nil).not_trashed.order(:position) }
+      has_many :trashed_elements,
+        -> { Element.trashed.order(:position) },
+        class_name: 'Alchemy::Element'
+      has_many :descendent_elements,
+        -> { order(:position).not_trashed },
+        class_name: 'Alchemy::Element'
+      has_many :contents, through: :elements
+      has_many :descendent_contents,
+        through: :descendent_elements,
+        class_name: 'Alchemy::Content',
+        source: :contents
+      has_and_belongs_to_many :to_be_swept_elements, -> { uniq },
```

Comment:
Rails/HasAndBelongsToMany: Prefer `has_many :through` to `has_and_belongs_to_many`.